src/types.jl

```diff
@@ -0,0 +1,171 @@
+
+## Hold a reference to a SymEngine object
+type Basic 
+    ptr::Ptr{Void}
+    function Basic()
+        z = new(C_NULL)
+        ccall((:basic_new_stack, :libsymengine), Void, (Ptr{Basic}, ), &z)
+        finalizer(z, basic_free)
+        return z
+    end
+end
+export Basic
+
+
+basic_free(b::Basic) = ccall((:basic_free_stack, :libsymengine), Void, (Ptr{Basic}, ), &b)
+
+
+function Basic(x::Clong)
+    a = Basic()
+    ccall((:integer_set_si, :libsymengine), Void, (Ptr{Basic}, Clong), &a, x)
+    return a
+end
+
+function Basic(x::Culong)
+    a = Basic()
+    ccall((:integer_set_ui, :libsymengine), Void, (Ptr{Basic}, Culong), &a, x)
+    return a
+end
+
+function Basic(x::BigInt)
+    a = Basic()
+    ccall((:integer_set_mpz, :libsymengine), Void, (Ptr{Basic}, Ptr{BigInt}), &a, &x)
+    return a
+end
+
+
+if Clong == Int32
+    convert(::Type{Basic}, x::Union{Int8, Int16, Int32}) = Basic(convert(Clong, x))
+    convert(::Type{Basic}, x::Union{UInt8, UInt16, UInt32}) = Basic(convert(Culong, x))
+else
+    convert(::Type{Basic}, x::Union{Int8, Int16, Int32, Int64}) = Basic(convert(Clong, x))
+    convert(::Type{Basic}, x::Union{UInt8, UInt16, UInt32, UInt64}) = Basic(convert(Culong, x))
+end
+convert(::Type{Basic}, x::Integer) = Basic(BigInt(x))
+convert(::Type{Basic}, x::Rational) = Basic(BasicType{Val{:Integer}}(num(x)) / BasicType{Val{:Integer}}(den(x)))
+
+
+
+## Wrapper type
+## this allows SymEngine.jl to keep track of the class of the C++ object
+## XXX This needs to be generated from symengine on startup
+## XXX This might be tedious with precompilation!
+const SYMENGINE_ENUM = Dict{Int, Symbol}(0 => :Integer,
+                                         1 => :Rational,
+                                         2 => :Complex,
+                                         3 => :ComplexDouble,
+                                         11 => :Symbol,
+                                         12 => :EmptySet,
+                                         13 => :Interval,
+                                         14 => :Mul,
+                                         15 => :Add,
+                                         16 => :Pow,
+                                         19 => :Constant,
+                                         20 => :Sin,
+                                         21 => :Cos
+                                         )
+
+## Parameterized type allowing or dispatch on Julia side by type of objecton SymEngine side
+## Use as BasicType{Val{:Integer}}(...)
+type BasicType{T} <: Number
+    x::Basic
+end
+
+Basic(x::BasicType) = x.x
+
+function get_type(s::Basic)
+    ccall((:basic_get_type, :libsymengine), UInt, (Ptr{Basic},), &s)
+end
+
+function get_class_id(id)
+    id = string(id)
+    ccall((:basic_get_class_id, :libsymengine), AbstractString, (Ptr{AbstractString},), &id)
```

Comment:
I see that and am waiting for it to be merged. In the mean time, I have
another attempt at this I'll push soon.

On Sun, Mar 27, 2016 at 11:35 AM, Isuru Fernando <notifications@github.com>
wrote:

> In src/types.jl
> <https://github.com/symengine/SymEngine.jl/pull/18#discussion_r57530336>:
>
> > +
> > +## Parameterized type allowing or dispatch on Julia side by type of objecton SymEngine side
> > +## Use as BasicType{Val{:Integer}}(...)
> > +type BasicType{T} <: Number
> > +    x::Basic
> > +end
> > +
> > +Basic(x::BasicType) = x.x
> > +
> > +function get_type(s::Basic)
> > +    ccall((:basic_get_type, :libsymengine), UInt, (Ptr{Basic},), &s)
> > +end
> > +
> > +function get_class_id(id)
> > +    id = string(id)
> > +    ccall((:basic_get_class_id, :libsymengine), AbstractString, (Ptr{AbstractString},), &id)
>
> I added a new method to get the class name from the id,
> basic_get_class_from_id
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/symengine/SymEngine.jl/pull/18/files/7dd53bb9569461346c056a3e6edfa50875b41b3c#r57530336>
>



-- 
John Verzani
Chair, Department of Mathematics
College of Staten Island, CUNY
verzani@math.csi.cuny.edu